## ***CNC 2024 MP***
## Partie I : PROGRAMMATION ET CALCUL SCIENTIFIQUE
### Produit d’une chaîne matricielle
#### A- Calcul du nombre minimum de multiplications dans le produit d’une chaîne matricielle
##### ***A.1- Algorithme naïf***

```
Fonction minProduit (T, i, j)
    Si i = j-1 Alors
        Retourner 0
    Fin Si
    R <-- liste vide
    Pour k = i+1 à j-1 faire
        c <-- minProduit(T, i, k) + minProduit(T, k, j) + T[i]*T[k]*T[j]
        Ajouter c à la liste R
    Fin Pour
    Retourner le minimum de R
```

In [25]:
def minProduit(T, i, j):
    if i == j - 1:
        return 0  # une seule matrice, pas de multiplication
    R = []
    for k in range(i + 1, j):
        c = minProduit(T, i, k) + minProduit(T, k, j) + T[i] * T[k] * T[j]
        R.append(c)
    return min(R)


In [26]:
T = [ 20, 10, 35, 8, 12 ]
minProduit (T, 0, len(T)-1) 

6160

##### ***A.2- Programmation dynamique Top-Down (Mémoïsation)***

In [27]:
def minPrd(T, i, j, D=None):
    if D is None:
        D = {}

    # Si déjà calculé, retourner la valeur stockée
    if (i, j) in D:
        return D[(i, j)]

    # Cas de base : une seule matrice => aucun coût
    if i == j - 1:
        return 0

    # Initialiser le coût minimal à l'infini
    min_cost = float('inf')

    # Tester toutes les coupures possibles
    for k in range(i + 1, j):
        cost_left = minPrd(T, i, k, D)
        cost_right = minPrd(T, k, j, D)
        cost_total = cost_left + cost_right + T[i] * T[k] * T[j]
        if cost_total < min_cost:
            min_cost = cost_total

    # Stocker et retourner le résultat
    D[(i, j)] = min_cost
    print(D)
    return min_cost


In [28]:
T = [ 20, 10, 35, 8, 12 ]
minPrd (T, 0, len(T)-1)

{(2, 4): 3360}
{(2, 4): 3360, (1, 3): 2800}
{(2, 4): 3360, (1, 3): 2800, (1, 4): 3760}
{(2, 4): 3360, (1, 3): 2800, (1, 4): 3760, (0, 2): 7000}
{(2, 4): 3360, (1, 3): 2800, (1, 4): 3760, (0, 2): 7000, (0, 3): 4400}
{(2, 4): 3360, (1, 3): 2800, (1, 4): 3760, (0, 2): 7000, (0, 3): 4400, (0, 4): 6160}


6160

Voici une **explication pas à pas** de l'exécution de la fonction `minPrd` pour calculer le **nombre minimal de multiplications** dans le produit d'une chaîne de matrices, en utilisant la **mémoïsation** via un dictionnaire `D`.

---

##### 1. **Initialisation**

- Soit `T = [20, 10, 35, 8, 12]` représentant les dimensions des matrices.
- On cherche à calculer le produit matriciel : `M0 * M1 * M2 * M3`, avec les dimensions suivantes :
  - `M0 = 20 × 10`
  - `M1 = 10 × 35`
  - `M2 = 35 × 8`
  - `M3 = 8 × 12`

###### Fonction `minPrd(T, i, j, D)` :
- `T` : liste des dimensions des matrices
- `(i, j)` : indices délimitant la sous-chaîne de matrices à traiter
- `D` : dictionnaire pour mémoriser les résultats déjà calculés (mémoïsation)

---

##### 2. **Appel initial** :

L'appel initial est :

```python
minPrd(T, 0, 4, D)
```

Cela correspond à calculer le produit des matrices de `M0` à `M3`. Le dictionnaire `D` est initialisé comme vide.

---

##### 3. **Première vérification** :

- Le dictionnaire `D` ne contient pas encore de résultats, donc on va effectuer les calculs.

---

##### 4. **Cas de base** :

- Si `i == j-1` : la fonction retourne 0 (une seule matrice, pas de multiplication).
  
Dans notre cas, ce n'est pas vrai, car nous avons plusieurs matrices à multiplier.

---

##### 5. **Exploration des sous-problèmes** :

On va tester toutes les coupures possibles en divisant la chaîne de matrices de différentes façons.

On parcourt `k` de `i+1` à `j-1`, ce qui permet de diviser la chaîne en deux sous-problèmes :
- Le produit de `M0...Mk` et
- Le produit de `Mk+1...Mn-1`.

###### Premier découpage : `k = 1`

- On découpe la chaîne en :
  - `minPrd(T, 0, 1)` (produit des matrices `M0` seule),
  - `minPrd(T, 1, 4)` (produit des matrices de `M1` à `M3`).
  
On fait ces appels récursifs et on continue à décomposer les sous-problèmes.

---

##### 6. **Calcul des sous-problèmes** :

On effectue des appels récursifs pour calculer les sous-problèmes suivants, avec la mémoïsation permettant de ne pas recalculer les mêmes résultats plusieurs fois :

- `minPrd(T, 0, 1)` : retourne 0 (une seule matrice, donc aucune multiplication),
- `minPrd(T, 1, 4)` : calculé récursivement comme suit :
  - On découpe `minPrd(T, 1, 2)` et `minPrd(T, 2, 4)`,
  - Ces appels récursifs calculent les sous-problèmes et stockent leurs résultats dans `D`.

---

##### 7. **Calcul du coût total** :

À chaque étape, on additionne les coûts des sous-problèmes et on ajoute le coût de multiplication des deux résultats intermédiaires :

- Lorsque `k = 1`, on calcule :
  
  ```
  minPrd(T, 0, 1) + minPrd(T, 1, 4) + T[0] * T[1] * T[4]
  ```

  Ce coût est ajouté dans la liste des coûts possibles.

- De même, pour d'autres valeurs de `k` (k = 2 et k = 3), on calcule les coûts des sous-problèmes.

---

##### 8. **Retour du coût minimal** :

À chaque niveau de la récursion, on compare les différents coûts calculés pour chaque coupure possible (`k`), et on garde le **minimum**.

---

##### 9. **Stockage et retour du résultat** :

Le résultat final est stocké dans `D[(0, 4)]` pour le produit complet de toutes les matrices de `M0` à `M3`, et il est retourné.

---

##### 10. **Exécution détaillée (dans l'ordre des appels)** :

Voici l’ordre d'exécution avec les appels récursifs :

```
minPrd(T, 0, 4, D)  -> (Découpe à k=1)
    -> minPrd(T, 0, 1, D)  # retourne 0
    -> minPrd(T, 1, 4, D)  -> (Découpe à k=2)
        -> minPrd(T, 1, 2, D)  # retourne 0
        -> minPrd(T, 2, 4, D)  -> (Découpe à k=3)
            -> minPrd(T, 2, 3, D)  # retourne 0
            -> minPrd(T, 3, 4, D)  # retourne 0
            # Calcule le coût pour k=3 et retourne
        # Calcule le coût pour k=2 et retourne
    # Calcule le coût pour k=1 et retourne
```

---

##### Conclusion

- Chaque appel récursif est mémorisé dans `D` pour éviter de recalculer les mêmes résultats.
- La fonction `minPrd` renvoie le **nombre minimal de multiplications** pour multiplier les matrices de `M0` à `M3`.
- Grâce à la **mémoïsation**, on réduit considérablement le temps de calcul en évitant les redondances.

---

***Explication étape par étape***
| Appel `minPrd(i, j)` | Valeurs de `k` testées | Calculs effectués                                                                 | Coût total             | Meilleur coût |
|----------------------|-------------------------|------------------------------------------------------------------------------------|------------------------|----------------|
| `minPrd(0, 4)`       | 1, 2, 3                 | (0 + 3760 + 2400), (7000 + 3360 + 8400), (4400 + 0 + 1920)                         | 6160, 18760, 6320      | ✅ 6160         |
| └ `minPrd(1, 4)`     | 2, 3                   | (0 + 3360 + 4200), (2800 + 0 + 960)                                                | 7560, 3760             | ✅ 3760         |
| └ `minPrd(2, 4)`     | 3                      | (0 + 0 + 3360)                                                                     | 3360                   | ✅ 3360         |
| └ `minPrd(1, 3)`     | 2                      | (0 + 0 + 2800)                                                                     | 2800                   | ✅ 2800         |
| └ `minPrd(0, 2)`     | 1                      | (0 + 0 + 7000)                                                                     | 7000                   | ✅ 7000         |
| └ `minPrd(0, 3)`     | 1, 2                   | (0 + 2800 + 1600), (7000 + 0 + 5600)                                               | 4400, 12600            | ✅ 4400         |


Pour calculer les éléments de la matrice C, on utilise l’algorithme (1) suivant :
![algo1](algo1.png)

In [17]:
def matriceC(T):
    n = len(T) - 1  # Nombre de matrices
    C = [[0 if i >= j else float('inf') for j in range(n)] for i in range(n)]

    # Longueur de la chaîne (de 2 à n)
    for l in range(2, n + 1):
        for i in range(n - l + 1):
            j = i + l - 1
            for k in range(i, j):
                cost = C[i][k] + C[k + 1][j] + T[i] * T[k + 1] * T[j + 1]
                if cost < C[i][j]:
                    C[i][j] = cost
    return C


In [18]:
matriceC(T)

[[0, 7000, 4400, 6160], [0, 0, 2800, 3760], [0, 0, 0, 3360], [0, 0, 0, 0]]

Voici l'explication de l'exécution de la fonction `matriceC(T)` étape par étape :


###### Fonction `matriceC(T)` - Explication

###### Description
La fonction `matriceC(T)` calcule le **coût minimal** pour multiplier une chaîne de matrices en utilisant la **programmation dynamique**. La fonction retourne une matrice `C` où chaque élément `C[i][j]` représente le **coût minimal** pour multiplier les matrices de `Mi` à `Mj`.

###### Entrée :
- **`T`** : Liste des dimensions des matrices.
    - Par exemple, `T = [20, 10, 35, 8, 12]` représente 4 matrices :
        - `M0` : 20 x 10
        - `M1` : 10 x 35
        - `M2` : 35 x 8
        - `M3` : 8 x 12

###### Sortie :
- **`C`** : Matrice carrée de taille `n x n`, où `n` est le nombre de matrices, et `C[i][j]` est le coût minimal pour multiplier les matrices de `Mi` à `Mj`.

---

###### Explication du code

###### Initialisation de la matrice `C` :

```python
C = [[0 if i >= j else float('inf') for j in range(n)] for i in range(n)]
```

- La matrice `C` est initialisée avec des valeurs infinies (`float('inf')`) pour les éléments où `i < j`, et `0` pour les éléments où `i = j`, car multiplier une seule matrice n'entraîne pas de coût.

---

###### Calcul des coûts minimaux pour les sous-chaînes :

La fonction parcourt différentes longueurs de chaînes, de 2 à `n`, et calcule le coût minimal pour chaque sous-problème.

```python
for l in range(2, n + 1):  # Longueur de la chaîne (de 2 à n)
    for i in range(n - l + 1):  # Début de la chaîne
        j = i + l - 1  # Fin de la chaîne
        for k in range(i, j):  # Position de la coupure
            cost = C[i][k] + C[k + 1][j] + T[i] * T[k + 1] * T[j + 1]
            if cost < C[i][j]:
                C[i][j] = cost
```

- **`l`** : Longueur de la chaîne de matrices.
- **`i` et `j`** : Définissent la sous-chaîne de matrices.
- **`k`** : Point de division de la chaîne de matrices pour calculer les coûts.
- **Coût** : Le coût est calculé par la formule :
  ```
  cost = C[i][k] + C[k+1][j] + T[i] * T[k+1] * T[j+1]
  ```

---

###### Exemple d'exécution

Prenons l'exemple où `T = [20, 10, 35, 8, 12]`.

###### Initialisation de `C` :
- Nombre de matrices `n = 4` (car il y a 5 éléments dans `T`).
- La matrice `C` est initialement remplie comme suit :
  ```
  C = [
      [0, inf, inf, inf],
      [0, 0, inf, inf],
      [0, 0, 0, inf],
      [0, 0, 0, 0]
  ]
  ```

###### Calcul des coûts pour des chaînes de longueur 2 :

- Pour `i = 0`, `j = 1` (chaîne `M0 * M1`), on n'a pas de coût car c'est une seule multiplication :
  ```
  C[0][1] = 0
  ```
  
- Pour `i = 1`, `j = 2` (chaîne `M1 * M2`), le coût est calculé comme suit :
  ```
  C[1][2] = 10 * 35 = 350
  ```

- Pour `i = 2`, `j = 3` (chaîne `M2 * M3`), le coût est :
  ```
  C[2][3] = 35 * 8 = 280
  ```

Matrice `C` après les calculs pour `l = 2` :
```
C = [
    [0, inf, inf, inf],
    [0, 0, 350, inf],
    [0, 0, 0, 280],
    [0, 0, 0, 0]
]
```

###### Calcul des coûts pour des chaînes de longueur 3 :

- Pour `i = 0`, `j = 2` (chaîne `M0 * M1 * M2`), on essaie différentes coupures `k` :
  - Pour `k = 0`, le coût est `inf`.
  - Pour `k = 1`, le coût est :
    ```
    C[0][1] + C[1][2] + 20 * 10 * 35 = 0 + 350 + 7000 = 7000
    ```

- Pour `i = 1`, `j = 3` (chaîne `M1 * M2 * M3`), le coût est :
  ```
  C[1][2] + C[2][3] + 10 * 35 * 8 = 350 + 280 + 2800 = 8400
  ```

Matrice `C` après les calculs pour `l = 3` :
```
C = [
    [0, inf, 7000, inf],
    [0, 0, 350, 8400],
    [0, 0, 0, 280],
    [0, 0, 0, 0]
]
```

###### Calcul des coûts pour des chaînes de longueur 4 :

- Pour `i = 0`, `j = 3` (chaîne complète `M0 * M1 * M2 * M3`), on essaie différentes coupures `k` :
  - Pour `k = 2`, le coût est :
    ```
    C[0][2] + C[2][3] + 20 * 35 * 12 = 7000 + 280 + 8400 = 11280
    ```

Matrice `C` finale :
```
C = [
    [0, inf, 7000, 11280],
    [0, 0, 350, 8400],
    [0, 0, 0, 280],
    [0, 0, 0, 0]
]
```

Le coût minimal pour multiplier toutes les matrices de `M0` à `M3` est `C[0][3] = 11280`.

---

###### Résumé

La fonction `matriceC(T)` utilise la programmation dynamique pour calculer le coût minimal de multiplication d'une chaîne de matrices. La matrice `C` contient les coûts minimaux pour chaque sous-problème, et le résultat final pour le produit complet de matrices est donné par `C[0][n-1]`.

- **Complexité temporelle** : `O(n³)` où `n` est le nombre de matrices.
- **Sortie** : La matrice `C` avec les coûts minimaux.


##### ***A.3- Programmation dynamique Bottom-Up***

### ***B- Recherche du « parenthésage » optimal***

![algo2](algo2.png)

In [21]:
def matriceP(T, C):
    n = len(T) - 1  # Nombre de matrices
    P = [[0 for _ in range(n)] for _ in range(n)]

    for l in range(2, n + 1):  # Longueurs des sous-chaînes
        for i in range(n - l + 1):
            j = i + l - 1
            for k in range(i, j):
                cost = C[i][k] + C[k + 1][j] + T[i] * T[k + 1] * T[j + 1]
                if cost == C[i][j]:
                    P[i][j] = k
                    break  # On a trouvé le bon k, inutile de continuer
    return P


In [24]:
T = [20, 10, 35, 8, 12]

# Matrice C obtenue avec matriceC (version de l'algorithme (1))
def matriceC(T):
    n = len(T) - 1
    C = [[0 if i >= j else float('inf') for j in range(n)] for i in range(n)]
    for l in range(2, n + 1):
        for i in range(n - l + 1):
            j = i + l - 1
            for k in range(i, j):
                cost = C[i][k] + C[k+1][j] + T[i] * T[k+1] * T[j+1]
                if cost < C[i][j]:
                    C[i][j] = cost
    return C

# Calcul
C = matriceC(T)
P = matriceP(T, C)

print("Matrice P :")
for row in P:
    print(row)

print("Matrice C :")
for row in C:
    print(row)


Matrice P :
[0, 0, 0, 0]
[0, 0, 1, 2]
[0, 0, 0, 2]
[0, 0, 0, 0]
Matrice C :
[0, 7000, 4400, 6160]
[0, 0, 2800, 3760]
[0, 0, 0, 3360]
[0, 0, 0, 0]


:smile:
